### DISCUSSION TITLE
Struggles with integrating 1Password's op

### DISCUSSION THREAD

**Discussion by rsyring** [jdx/mise#3542] [Troubleshooting and bug reports] [answered]
I've been trying to figure out how to get 1Password secrets integrated with mise (refs #1359).  The biggest challenge I've encountered is that the `op` utility is rather slow, taking about 1s to return when asked for secrets.  That delay on every prompt is pretty annoying.

Since direnv doesn't run at every prompt, I thought I had this solved with a pretty simple setup:

```toml
# .mise.toml
[env]
TF_VAR_wasabi_access_key = 'op://private/aws/access-key'
TF_VAR_wasabi_secret_key = 'op://private/aws/secret-key'
```

```sh
# .envrc
direnv_load op run --no-masking -- direnv dump
watch_file .mise.toml
```

This setup has the advantage of being usable with `op run` directly if someone prefers not to use direnv.  The problem I encountered is that there is a conflict between direnv and mise which led to the environment variables not getting removed when leaving the directory.  The secrets were no longer resolved but the environment variables (`TF_VAR_*`) remained in the current shell.

The benefit of direnv for this setup is that `.envrc` isn't executed on every prompt.   If I need to refresh the secrets its easy with `direnv reload`.

Since mise has the stated intention of replacing direnv for any use case, it would be better to do:

```
# .mise.toml
[env]
TF_VAR_wasabi_access_key = 'op://private/aws/access-key'
TF_VAR_wasabi_secret_key = 'op://private/aws/secret-key'
_.source = 'op-run.sh'
```

But, in this case, for performance to be acceptable, I'd only want `op-run.sh` run the first time it was needed and/or on demand.

This may also also impact #1448.

One other thought regarding this: since we are dealing with secrets, I wouldn't want to see these values cached to the filesystem.

**Comment by andrewthauer** [comment]
I'm also using `direnv` with `op` to do similar things atm so I would find this useful.

I'm curious if there is a way to opt of out having the `_mise_hook` a `precmd_function`? I think if there was a way to activate `mise` as only a `chpwd_function`, this might resolve your issue. The alternative might be to avoid activating mise on shell startup and use `eval $(mise hook-env)` instead on demand.

**Comment by jdx** [comment]
you could experiment by running `mise activate zsh > some_file` then sourcing it in your zshrc with modifications

**Comment by jdx** [comment]
I'm not totally sure how we could resolve this since that would only work for zsh. Maybe what we could do is reuse previously resolved env vars if they've already been resolved and maybe offer a `mise refresh` command or something to get it to re-resolve them.

Gratefully, this would be feasible now that the `env._` changes have been made since that logic is centralized and not part of config loading anymore.

**Comment by rsyring** [comment]
> Maybe what we could do is reuse previously resolved env vars if they've already been resolved and maybe offer a mise refresh command or something to get it to re-resolve them.

This is exactly how it works (for me anyway) with direnv.  If I have to adjust env vars after changing directories into the project, then I run `direnv reload`.

I'm assuming this would apply to scripts too?  If so, I could see the possibility that some scripts people are running now are intended to be ran on every prompt.

> Gratefully, this would be feasible now that the env._ changes have been made since that logic is centralized and not part of config loading anymore.

:)

**Comment by rsyring** [comment]
> Maybe what we could do is reuse previously resolved env vars if they've already been resolved and maybe offer a mise refresh command or something to get it to re-resolve them.

Because these are secrets, I wouldn't want them cached to the file system.  That means they'd have to stay in the shell.  That would be a pretty big shift in architecture to what mise does now?

**Reply by airtonix** [reply]
But putting them in environment variables is any better?

**Reply by jkp** [reply]
The 'op run --' wrapper solution seems the cleanest option on the table - it lets the user scope sensitive info to the spots that need it and doesn't require changes in mise

**Reply by airtonix** [reply]
My point is that largely, secrets should (if possible) never be put into environment variables.

I don't care what the twelve factor cargo cult says.

So i really think this is an anti pattern that shouldn't be supported.

**Reply by rsyring** [reply]
So what's your solution then?  If not environment variables and not command line then you'd need some kind of universal support for "get your credentials this way" so each application/process could request them from some kind of credentials helper process.  The helper would then, of course, need to know the process was permitted.  Hows that all going to work?

I think there is a reason we don't have anything better than environment variables.  It would be great if there was something better.

The problem with only using `op run` is that it's dog slow.  If you have something like Terraform or Ansible where you have multiple credentials that need to be loaded you are talking an extra 15-20s startup time.  

Loading the environment variables into the shell doesn't make them available in the process's env vars.  They are going to be available to every command that you run in that shell, which is why env-config can be scoped to use only the ones you need for any given shell.

The benefit of env vars over plain text files in this case is that you can't accidentally include the file in a plain text archive or backup or accidentally commit and push to GitHub.

**Reply by reitzig** [reply]
Also, env vars are not accessible to other processes run by the same user (afaik), other than files.

> The problem with only using op run is that it's dog slow. If you have something like Terraform or Ansible where you have multiple credentials that need to be loaded you are talking an extra 15-20s startup time.

Does the time overhead really scale with the number of secrets? :thinking: I haven't done rigorous testing, but my impression is that there's an initial overhead (esp. if you need to unlock the desktop app), but subsequent accesses in the same `op` session are fast(er).

Also, you'd have that same overhead when loading/filling env vars -- unless you're saying you leave an authenticated shell open for long times? 😲

**Reply by rsyring** [reply]
> Also, env vars are not accessible to other processes run by the same user (afaik), other than files.

They **are** readable by other processes of same user:

```
 ❯ ps aux | grep cups-browsed
cups-br+  129680  0.0  0.0 498440 19192 ?        Ssl  Jun19   0:00 /usr/sbin/cups-browsed


 ❯ cat /proc/129680/environ | tr '\0' '\n' | head -n 1
cat: /proc/129680/environ: Permission denied


 ❯ ps aux | grep slack | head -n 1
rsyring     8834  0.0  0.3 1459714984 214308 ?   SLl  Jun18   0:53 /snap/slack/196/usr/lib/slack/slack --no-sandbox --executed-from=/home/rsyring --pid=8728


 ❯ cat /proc/8834/environ | tr '\0' '\n' | head -n 1
SHELL=/bin/bash
```

> Does the time overhead really scale with the number of secrets?

It does.  I just had an Ansible integration that did a number of direct calls to `op read` and that's where I got the 15-20s startup time from.

> Also, you'd have that same overhead when loading/filling env vars -- unless you're saying you leave an authenticated shell open for long times?  😲

That's exactly what I do.  :)  The shell stays open for as long as I'm working on that particular project.  But, the way I use my shells, is that I have tabs open for each individual tool I'm using.  I'f I'm working on Project A, then I have a shell window for Project A with tabs open for Terraform with only it's creds, another for Ansible with only it's creds, and maybe a few other ones if/when needed which may have creds depending on what I'm doing in that shell.  If I'm done with a tool, then I close it's tab.  When I'm done with Proj A, I close the Window and all remaining tabs/shells (and their env var secrets) go away.

I never use those shells for misc work or other projects.  

Security is has trade-offs.  This seems to be a reasonable one for me.  :)

**Reply by reitzig** [reply]
> They are readable by other processes of same user

TIL, thanks!

To me, that means I should _definitely_ not have processes with authentication tokens in their environment lying around. 😅 That script that looks for, say, `AWS_SECRET_ACCESS_KEY` in all accessible `/proc/*/environ` seems a little _too_ simple for my taste.

**Reply by ncadou** [reply]
> My point is that largely, secrets should (if possible) never be put into environment variables.
> 
> I don't care what the twelve factor cargo cult says.
> 
> So i really think this is an anti pattern that shouldn't be supported.

While I like the idea, sometimes you work with software that you don't have control over and requires secrets in env vars. Extracting those secrets from whichever secure store can take many seconds in my case, which I can't tolerate happening on every single prompt. Now the question is, is there a tool that can help alleviate that pain? Even though mise promises to be a direnv replacement, this particular issue makes it unable to replace direnv in this scenario. So I have to stick with direnv, which then triggers mise. I do wish I could use a single tool, as I think mise is pretty nice otherwise.

**Reply by airtonix** [reply]
> So what's your solution then? 

Your code that you own should be using an in process pipeline to your secret store.

It's a common pattern but often overlooked because of the tutorial-driven-development nature the software industry has taken over the last ten years. (it's not a concept that makes most tutorials easily digestible)

so a fetch on demand and cache in process memory. 

examples: 

- https://middy.js.org/docs/middlewares/ssm/
- or just roll your own es6 Proxy zod enriched object that uses aws-ssm client then falls back to env vars.

**Comment by jdx** [comment]
I'm not talking about caching to the file system, this would go into a different internal environment var like __MISE_DIFF. I think we'd need to track all of the current exported variables and what their source is—since if 2 config files export the same env var key that would require invalidation.

> This is exactly how it works (for me anyway) with direnv.

Good to know, it's probably safe for us to do similar logic.

**Comment by jdx** [comment]
There shouldn't be any security risk here since the env vars already contain the secret's value. In fact we don't even need to track the value of the keys I don't think since they're already exported, we just need to track their source

**Comment by rsyring** [comment]
> There shouldn't be any security risk here since the env vars already contain the secret's value. In fact we don't even need to track the value of the keys I don't think since they're already exported, we just need to track their source

Good to know.  I didn't realize the shell environment variables were persistent between prompts.

**Comment by andrewthauer** [comment]
@rsyring - Correct me if I'm wrong, but wasn't your main concern around invoking a script that in turn triggers a 1password secret dump on each prompt/cmd?

I'm currently using `op` with `direnv` in some projects to trigger a script to set the secrets on cd change into the root project directory (via direnv). This then prompts for biometrics auth on every shell opening. It doesn't however prompt for auth on cd change within the project so it's not horrible.

Even on entry to the root can be a bit much and I've experimented and often just use a shell function ondemand when I need the secrets loaded, since many commands don't require them.

However, my point of this is that if we rely on the current `_mise_hook` to run `mise hook-env` this will occur on every cmd prompt (not just cd)? I don't think 1password will prompt each time for biometrics auth, but it will take a significant performance hit to run a source script that does the secrets dump.

So, I think the only reasonable approach would be to only do this on cd change into the project root. Back to my question about only running the `_mise_hook` on `chpwd` rather then `precmd`.

**Comment by rsyring** [comment]
@andrewthauer I may not be fully understanding your comments.  The point of this issue was to address being able to run a script once (presumably on changing into a project directory) instead of at every prompt.  I think you are saying the same thing.  

We've also discussed just above being able to load on demand (e.g. `mise reload`, like `direnv reload`).

I guess another consideration would be to have the option to ONLY load on demand.    For some of my projects, where I'm managing infrastructure, I always need secrets.  For a others, I almost never do because I can do most of the development locally and only occasionally want to run against a live service.  e.g. `mise exec --keep-env secrets.sh`

**Comment by andrewthauer** [comment]
Essentially, yes, I'm saying the same thing, but I guess I'm getting into the the technical details of how `mise` is working atm. When you `eval $mise activate $shell)` this is creating a hook that essentially calls `mise hook-env` on every cd change + every comment prompt. If you have something like `env._.source` which in turn is executing `op` to get secrets, this is going to be executed every command prompt since the prompt/cd change hook is calling `mise hook-env`.

So not activating and manually calling `eval $(mise hook-env $shell)` once on demand would avoid the extra cost of invoking the sourced script each prompt/cmd. However, you lose the benefit of mise auto activate in this case (if you care about that workflow).

What I was ultimately trying to say, which probably did a horrible job at, was that you can't really get your cake and eat it while using the mise activate in your zshrc, bashprofile without getting the performance hit on each prompt. Which isn't noticeable normally but would be for this sort of use case. However, if there was a way to say I only want to use the cd change hook (chpwd in zsh speak), then this hit would only occur on a cd change.

I'm not sure how mise would selectively only run the `source` script in some cases and not in others. For example if you `direnv` into a project directory with an `.envrc` file it will execute that. If that calls to `op` it runs it. However, if then you cd into a child directory it won't re-execute the `.envrc` file in the parent directory. Not sure this is the case with `mise` atm, but imo this would be desirable.

**Comment by jdx** [comment]
> When you eval $mise activate $shell) this is creating a hook that essentially calls mise hook-env on every cd change + every comment prompt.

while accurate there is some [short circuiting logic](https://github.com/jdx/mise/blob/main/src/cli/hook_env.rs#L39) that might not be working correctly here. This should be preventing `env._.source` from running every time unless something isn't working right.

**Comment by andrewthauer** [comment]
TBH, I haven't tested this particular use case with `mise` yet. So if there is short circuiting logic then that might suffice along with the suggested on demand reload ability.

**Comment by andrewthauer** [comment]
So I had a go at porting my current `op` + `direnv` setup to `mise`. I think it can work as is, but ... there are at least a couple of major issues that make it not feasible atm (at least for me).

First I'll document my 2 current approaches for completeness.

**Method 1: Direnv library helper**

This combines a `direnv` re-usable function that uses the `direnv_load` & `direnv dump` which @rsyring is also using. It works relatively well, since when I cd into a project I get prompted by 1password for auth to resolve the secrets in the .env file that contains secrets refs then makes them available in my current shell. Subsequent command within the project are not slow as it doesn't seem to attempt to re-run the `use_op_env` in my `.envrc` file.

```sh
# ~/.config/direnv/lib/op.sh
use_op_env() {
   # implementation omitted to be brief, but essentially does a `direnv_load op run -- direnv dump`
}

# Then in a projects .envrc file
# ~/src/some-project/.envrc
...
use op_env some-account.1password.com
```

**Method 2: On demand shell alias**

This works well for either some shared global secrets and does not rely on direnv at all. It' also avoids getting the potentially annoying upfront prompt from 1password for auth or startup hit.

```sh
# ~/.local/bin/op-dotenv
# this essentially replaces the direnv_load part and does:
#   `cat <"$dotenv_file" | op inject` and then spits out the resulting resolved variables

# ~/.zshrc
alias ope='eval $(op-dotenv "$HOME"/.config/.global-env --export)'

# then at any point run this alias to get those secrets into your shell. op will prompt for auth the first time in the shell
ope
```

**Usage in `mise`**

In `mise` since I want to avoid usages of `direnv` I needed to adapt method 2 and came up with this ...

First the file to source which in turn calls the `op-dotenv` helper which invokes `op` and spits out the resolved env variables with `export` in front.

```sh
# ~/src/some-project/.config/scripts/op-env.sh
op-dotenv "$PWD"/.env --export
```

Then in the mise config

```toml
# ~/src/some-project/.mise.toml
[env]
_.source = "./config/scripts/op_env.sh"   # file above
```

NOTE: My setup could be adapted likely to work with variables defined in `[env]`. The `op-dotenv` would likely end up need to be run after the `[env]` variables are applied, then do something like `env | op-dotenv` which would in turn resolved any secret references in the env and export the resolved versions on top.

This works as expected, however, 2 issues I ran into:

1. It seems to invoke the sourced `op_env.sh` script after every command is executed. So, I think this confirms the short circuiting may not be working?

2. If my underlying `op-dotenv` executable script uses something that is managed by `mise` it gets caught in and endless loop. For example I have 2 versions of the `op-dotenv` I wrote. One in `bash` and one in `deno`. The bash one works fine, but the deno one infinity loops. My suspicion is this is because I use `mise` to manage my `deno` versions. I can file a separate issue for this.

So, in the end it seems definitely doable, but the 2 points above would need to be addressed before it can be widely used I think. The other thing that would be nice is a way to used "share" helpers like direnv through some common place to look up plugins (e.g. `~/.local/config/mise/lib`).

**Comment by jdx** [comment]
I know this wouldn't be perfect but I wonder if somehow using the enter hook would solve this problem. You could have that in your global mise.toml which would effectively only run once. That can't export env vars though so idk how helpful it would be. Perhaps we could extend that somehow to allow for this.

Actually this made me realize that there is a bug using the global config, it never gets run for enter, but that's a seperate problem.

**Comment by jdx** [comment]
the specific issue where mise doesn't run if config files didn't change should be fixed now that hooks are out. As part of that I needed to make hook-env run anytime the directory changes.

**Comment by andrewthauer** [comment]
I tried revisiting this setup and got a little further. I tried 2 approaches:

1. using `_.source = "some-script.sh` that resolves secrets from 1password CLI, and then `exports` the resulting env vars. For some reason the exported env vars where not showing up in my parent shell, but I did get prompted by 1password. Would need to debug this more.

2. I created a quick POC of a lua plugin with a `mise_env.lua` hook, that calls out to 1password CLI and then modifies the path. This works and the resolve env vars show up in the parent shell.

However, both of these solutions currently suffer from the fact they get executed every single prompt (even non mise ones). I think twice actually, but not sure. This means it's a no go since the secrets resolution can take a few seconds. 

With my current `direnv` solution it only happens the first `cd` into the directory containing the `.envrc` file. I can cd into child directories, etc. run commands and it won't try and re-run the file. I haven't looked, but I suspect direnv does some hashing of the current state and the .envrc file to determine if it needs to run or not when changing directories. Imo, something like this would be a hard requirement for me being able to use `mise` for the purposes of loading secrets or running longer processes to setup the env for a project.

@jdx - Am I missing anything currently built in that would help here? Or is this something new that would need to be incorporated into `mise`?

**Comment by jdx** [comment]
> I think twice actually, but not sure.

this is likely shell specific but I did recently discover when changing directories we effectively call twice (once for cd and once for the prompt). I'm not sure I actually _want_ to fix that because that code inevitably causes unexpected bugs whenever I touch it. `hook-env` should be fast on the second execution due to early exit though

I did recently add support for [shell hooks](https://mise.jdx.dev/hooks.html#shell-hooks) which would be a partial, albeit ugly way to solve this. The problem with that is this is straight up executing shell code—mise won't be able to remove env vars added there once you leave. While you could in theory write a `leave` hook to manually clean up, I haven't implemented that hook yet and it's not trivial to add that one because reasons.

Ultimately it sounds like what you're looking for is the ability to use `_.source` (assuming it worked correctly) but with the ability to only run it on `cd` and not every prompt. btw there is an [e2e test](https://github.com/jdx/mise/blob/8f40147972606bd45f3696f5ea50feeeb238bff0/e2e/cli/test_env_source#L5) for this so it should be working. Maybe this could be configured like this:

```toml
[env]
_.source = {path = "some-script.sh", when="cd"}
```

As far as implementation, this would be a bit tricky since typically re-resolve this every time so there would need to be some caching in place. This would either be in the form of caching as a file (which is probably easier-implementation wise since we do similar things) or storing previous values in a distinct env var (as serialized data) like `__MISE_ENV_SOURCE` or something.

Alternatively, we could extend this to the hooks functionality as something like this:

```toml
[hooks]
enter = {source = "some-script.sh"}
```

Though I think this would be harder, since the hooks would somehow need to communicate the env var change in a way that it gets reset later. I think I don't like this idea.

**Reply by ncadou** [reply]
> ```toml
> [env]
> _.source = {path = "some-script.sh", when="cd"}
> ```

Interesting. This sourcing-at-every-prompt prevents me from switching to mise, as some projects I work on pull too many secrets from password-store to be practical with mise, and I was about to post my own suggestion when I found this discussion. Although maybe less of a concern, to fully emulate how smoothly direnv works in my case, I was also thinking about hot-reloading when changes are detected:

```toml
[env]
_.source = {path = "some-script.sh", when="cd", watch = true}
```

**Reply by syhol** [reply]
Just in-case you didn't know the current state of things @ncadou, the current recommended way of working with [secrets](https://mise.jdx.dev/environments/secrets.html) in a password store is to use the password store cli to create a local json/yaml/toml file, then use sops/age to encrypt it. mise can read these sops/age files **very fast** with the file directive:
```toml
[env]
_.file = ".env.json"
```
You could automate this with a few lines in a [mise task](https://mise.jdx.dev/tasks/), and (if you wanted to) you could regenerate this file in an [enter hook](https://mise.jdx.dev/hooks.html#enter-hook):
```toml
[env]
_.file = ".env.json"

[hooks]
enter = "mise run fetch-secrets"

[tasks.fetch-secrets]
run = '''
dcli inject --in .env.json.template --out .env.json
sops encrypt -i --age "~/.config/mise/age.txt" .env.json
'''
```
This is with the dashlane cli and you can imagine the .env.json.template looks something like:
```json
{
  "AWS_ACCESS_KEY_ID": "{{ dl://7593DD31-5FA5-40F4-85B8-F5131FB6C617/password }}",
  "AWS_SECRET_ACCESS_KEY": "{{ dl://88C50F1F-E779-4DBE-92CD-63713F00E653/password }}"
}
```

**Reply by dkarter** [reply]
@jdx I believe the best solution would be something like this:

```toml
[env]
_.source = {path = "some-script.sh", when="enter"}
```

Would that be possible?

What I imagine is, if you are entering a dir, and it has a `mise.toml`, and it has that directive, only then execute the script, then when leaving the dir clean it just as before (unless it's a sub directory)

**Reply by jdx** [reply]
my inclination would be no since it would only work on enter when using `mise activate` but not `mise x|run|env`, shims or any other way to use mise—unless we started putting state into how things like `mise env` works but I would be even less in favor of that. Even though `[hooks]` has some similar caveats around how you have to use mise I think it's simpler. I'd be more in favor of enhancing `[hooks]` which is still experimental and I think has some room to grow.

**Reply by ceelian** [reply]
@syhol The problem I see with this idea from a threat modeling perspective is that, as of now, the private key used for decryption is stored as a file at ~/.config/mise/age.txt. Therefore, any tool/person that has access to the encrypted env file and the private key file has access to the secrets. That's not much better than having the secrets unencrypted and gitignored. The advantage we get from a password manager like 1Password or Dashlane is that it can only give these secrets based on certain multi-factor requirements and other corporate policies.

To build on your idea with sops, I would suggest having a setting in mise.toml that can call a script/command that returns the private key for sops to stdout. That way, the private key for decryption is never stored on disk — just in memory — and we could execute the CLI of the password manager to retrieve the private key for sops only, e.g., after providing multi-factor authentication.

But I'm not sure if that approach won't re-introduce the original problem that mise must execute that script that retrieves the private key so often that we again have that performance issue.

**Reply by ceelian** [reply]
The best workaround/solution I could come up with, without having to save the secrets to a file, is by using shell hooks. Maybe I am missing some more disadvantages of this approach, but my demo works as expected. The only disadvantage I found, and as @jdx mentioned, is that hooks run outside the environment system of mise, so with hooks we need to clean up the set env variables ourselves. But that's a caveat that is okay for me for the moment.

My toml looks like this (stripped unimportant parts):

```
[env]
_.file = "./current.env"
_.source = "config/{{env.MISE_ENV}}.sh"

[hooks.enter]
shell = "zsh"
script = "source ./config/{{env.MISE_ENV}}.secrets.sh"

[hooks.leave]
shell = "zsh"
script = "source $OLDPWD/config/{{env.MISE_ENV}}.secrets.sh unset"
```

In the `prod.secrets.sh` I call the 1password cli and read the secrets to env variables that I export. In the leave hook, we need to unset the env variables for security reasons. Unfortunately, it looks like no MISE_PROJECT_ROOT env variable is set in the leave hook, so we need to use the $OLDPWD variable to create the path to the script for unsetting the variables.

Any feedback on this approach is highly appreciated — maybe I overlook some potential issues here.

**Reply by clayrosenthal** [reply]
@ceelian I did something similar, but instead of `OLDPWD` I used `{{ config_root }}` assuming that the `/config` path is relative to the same dir as your `mise.toml`

**Reply by jarv** [reply]
On disadvantage to be aware of using `hooks.enter` and `hooks.leave` to set/unset variables that if you have any variables with name conflicts set in your `secrets.sh`, the original values won't be restored when you leave the project directory.

**Comment by rsyring** [comment, marked as answer]
FWIW, I ended up realizing that setting environment variables was not always straight forward for a project.  A decent number of my projects actually had groups of environment variables that needed to be set based on what I was trying to do.  Different cloud environments, running Ansible vs Terraform for a project, etc. etc.  That was true for more than just secrets as well, maybe API URLs changed, maybe I wanted to set `AWS_PROFILE` differently.

I also realized I often didn't want secrets loaded whenever I was in a project's directly b/c I wasn't going to do anything that used a secret.  I.e. I was just going to be doing local development and would not be deploying anything.

I wanted to have the ability to define these profiles as well as integrate with secrets (we use 1Password) so I created [env-config](https://github.com/level12/env-config). 

We still use mise for a lot including static environment variables but now use env-config for secrets and vars that are context/activity dependent.

**Reply by jdx** [reply]
makes sense to me, trying to put an extra process in the critical path of mise loading the environment is likely always going to result in poor performance to some extent

**Comment by reitzig** [comment]
FWIW, instead of loading secrets into the environment, we can also use `op run` to wrap the commands that actually need the secrets, or use `op read` in scripts. That considerably narrows the time frame when secrets are exposed, and it's also rather convenient to use (IMHO). For repetitive use, one could wrap the calls in mise tasks.

**Reply by jkp** [reply]
This!

**Comment by 134130** [comment]
Running the op-cli every time you change a directory is too slow and cumbersome, especially if you're setting up 2FA.

Many projects define an `init` task to set initial environment variables.

You can see how I use init in my mise task [here](https://github.com/134130/mise-en-monorepo/blob/6b87c50bc0799ff892d7fc7d602aa38ccc32db0c/mise.toml#L5-L56). This is a way to utilize mise's local profile.

**Comment by dudil** [comment]
@reitzig @134130  loved your concept with the wrapping ```op run``` and came up an elegant solution that leverages mise's config layering for automatic secret injection. This is more useful for me since we use 1password only on local machine setups and use GitHub secrets when running CI/CD.

@rsyring , @jkp - I hope you would find it useful as well.

### setup
``` toml
# mise.toml
# --- more config above
[tasks.base]
description = """
    This is a base task for overriding commands start in mise.<env>.toml
    It sets up the environment by injecting secrets, for example, when in local environments.
    """
run = ""
# --- more config below
```

```toml
# mise.local.toml
# --- more config above
[tasks.base]
tools = { "op" = "latest" }
run = 'exec op run -- "$@"'
# --- more config below
```

So basically the ```mise.local.toml``` base override the one declare at ```mise.toml```

### usage
``` toml
# mise.local.toml
# --- more config above
[tasks.plan]
run = "mise run base terraform plan"

[tasks.deploy]
run = "mise run base terraform apply"
# -- more config below
```
### Benefits
* DRY: All tasks use the same pattern
* Environment-aware: Secrets from 1Password only in local dev, regular execution in CI/CD
* Performance: process-scoped secrets (no persistent env vars)
* Zero config oeverhead in production

The **```base```** task becomes your secret injection point locally but remain transparent in CI/CD. Tasks works identically everywhere while maintaining security best practices (avoid polluting your shell with secrets, use the as JIT).

**Reply by esacteksab** [reply]
This is leveraging tasks, this doesn't necessarily work in a `direnv` like workflow where `cd` to a directory instantiates the environment variables? I guess I could wrap the tasks that need these env vars in a task, but not a seamless experience.

**Reply by dudil** [reply]
Got you and agree that YMMV.
From my (very short and pleasant) experience with mise, mise tasks are extremely powerful tool and I believe this is one of the killer features in mise! (Thank you for that @jdx ❤️ )
Yes - they work great with CD, since mise always scan the repo folders up to check for mise.*.toml files, you don't need to do anything special for that - it is just working!

IMHO, I can say that having secrets at env on every CD is highly risked and I would prefer to give up a total seamless experience for the sake of better security. 

My team is using NX, and just few days ago they had enormous breach that targeted secrets from process.env:
https://www.stepsecurity.io/blog/supply-chain-security-alert-popular-nx-build-system-package-compromised-with-data-stealing-malware

So needles to say I'm happy we are using mise tasks and not injecting secrets to direnv like workflows 🙏

P.S - I'm currently working with terragrunt and it makes using tasks even easier to use with IaC.
If you would like I can share with you the snippet

**Reply by esacteksab** [reply]
I would definitely love to see how you're using Terragrunt. I'm just using vanilla Terraform, but I do have remote state in S3 buckets and `direnv` uses `aws-vault` to auth against AWS so I can use the remote state. Thanks!

**Reply by dudil** [reply]
Sure:

My units are inside infra dir

``` toml
# mise.toml

# this one is consolidated from the actual env
# see pervious example on how to decouple it for cicd
[tasks.base]
tools = { "op" = "2.30" }
run = 'exec op run -- "$@"'

[tasks.tera]
description = "Run Terragrunt commands"
env = { UNIT = "--all" }
run = """
COMMAND={{arg(name='command')}}
UNIT={{arg(name='unit', default='--all')}}
if [ "$UNIT" != "--all" ]; then
    UNIT="--working-dir {{config_root}}/infra/${UNIT}"
fi
mise run base terragrunt ${COMMAND} ${UNIT}
"""

# The tasks below are just there to save me few clicks each time
# They are extremely optional 😅
[tasks.tr-init]
description = "Initialize Terragrunt configuration"
run         = "mise tera init {{arg(name='unit', default='--all')}}"

[tasks.plan]
description = "Plan Terragrunt configuration"
run         = "mise tera plan {{arg(name='unit', default='--all')}}"

[tasks.apply]
description = "Apply Terragrunt configuration"
run         = "mise tera apply {{arg(name='unit')}}"

[tasks.destroy]
confirm     = "true" # if you want to be extra careful
description = "Destroy Terragrunt configuration"
run         = "mise tera destroy {{arg(name='unit')}}"
```

So if I want to run terragrunt plan I could either:
``` bash
# Will run: terragrunt plan --all, injecting secrets only for this process
mise plan
# Will run terragrunt plan --working-dir ./infra/storage
mise plan storage
```

**Comment by kujenga** [comment]
In case it is helpful to anyone else, I created https://github.com/kujenga/mise-env-1password based on https://github.com/jdx/mise-env-fnox for a way to plug the `op` 1password CLI directly into mise rather than needing an intermediary through `fnox`.

**Reply by rsyring** [reply]
Cool project, thanks for the work and for sharing.

**Comment by zeitlinger** [comment]
**Cached `_.source` approach — single `op` call + tmpfs cache**

I ran into the same issue: putting `op read` calls in `[env]` via templates made every directory change slow (~12s for 3 secrets).

The fix is `_.source` with a caching wrapper that calls `op inject` once and caches the result in tmpfs (RAM-only, never hits disk, gone on reboot).

**`~/.config/mise/secrets.env.tpl`** — template for `op inject`:
```
GITHUB_TOKEN=op://Private/GitHub Token/credential
ANTHROPIC_API_KEY=op://Private/Anthropic Key/credential
```

**`~/.config/mise/load-secrets.sh`** — cached loader:
```bash
#!/bin/bash
# Cache in tmpfs — never written to disk, gone on reboot.
CACHE="/run/user/$(id -u)/mise-secrets.env"
TEMPLATE="$HOME/.config/mise/secrets.env.tpl"
MAX_AGE=3600

_needs_refresh=true
if [[ -f "$CACHE" ]]; then
  age=$(( $(date +%s) - $(stat -c %Y "$CACHE") ))
  (( age < MAX_AGE )) && _needs_refresh=false
fi

if $_needs_refresh; then
  if output=$(op inject -i "$TEMPLATE" 2>/dev/null); then
    : > "$CACHE"
    while IFS="=" read -r key value; do
      [[ -n "$key" && "$key" != \#* ]] && echo "export $key=\"$value\"" >> "$CACHE"
    done <<< "$output"
    chmod 600 "$CACHE"
  fi
fi

[[ -f "$CACHE" ]] && source "$CACHE"
```

**`~/.config/mise/config.toml`**:
```toml
[env]
_.source = { path = "~/.config/mise/load-secrets.sh", redact = true }
```

**Result**: First run after cache expires takes ~12s (includes biometric prompt). Every subsequent activation: **~6ms**.

The cache lives in `/run/user/$UID` which is tmpfs on Linux — secrets stay in RAM only, never touch disk, and disappear on reboot. Falls back gracefully if `op` is unavailable (locked/offline). `redact = true` keeps secrets out of `mise env` output.

**Comment by ivannovazzi** [comment]
The 1s latency on `op` is a known pain point — it basically calls out to the 1Password service for every invocation.

One approach that avoids that: use a secrets manager designed for CLI workflows where the injection is designed to be fast. [KeyEnv](https://keyenv.dev) is worth a look if your main concern is team secrets (API keys, tokens, etc.) rather than password management — the CLI injection step is much lighter weight than `op`.

That said, if you're committed to 1Password for passwords+secrets, caching the result locally (write the value to a temp file, check mtime) is a common workaround for the latency issue.

**Reply by syhol** [reply]
If only the creator of mise also created a secrets manager...

https://github.com/jdx/fnox

😮

**Comment by theoephraim** [comment]
The latency here is because `op read`/`op run` make several network roundtrips to 1Password's API on every invocation, even though all of your vault data is already sitting (encrypted) on disk in the desktop app. So any approach that resolves on each prompt pays that cost every time, which is exactly what's biting you. Until they fix it, caching is the only answer. Problem is that **you need a secure place to store that cache data**

Over at [varlock](https://varlock.dev), we have a full env toolkit with many plugins, including our [1Password plugin](https://varlock.dev/plugins/1password) which is what we use ourselves. We just rolled out [built-in caching](https://varlock.dev/guides/caching/) which is secured by secure enclave on mac, windows hello, TPM, etc. Varlock includes small native binaries to facilitate this and to manage "sessions", so you are not scanning your finger on every read. We have similar "session" logic as `op`, so individual terminal sessions must each be unlocked, although we are a bit more flexible about determining cases like turbo, or LLM sessions in code/claude desktop - cases where currently `op` would prompt you every time.

Varlock also provides built-in validation, type-safety, drop-in integrations for many frameworks, imports, 16 plugins, tons of other cool features...

Anyway one of our top priorities is making the DX of using 1Password better for dev-y things - so please check it out and pop in our discord if you run into any trouble.

A deep dive into nicer mise-specific integration is on my to-do list, but I think you should probably be able to get things going.
